pi5/pinctrl_utils.go

```diff
@@ -286,8 +289,88 @@ func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+/*
+For all Pins belonging to the same bank, pin data is stored contiguously and in 8 byte chunks.
+For a given pin, this method determines:
+ 1. which bank the pin belongs to
+ 2. the starting address of its 8 byte data chunk
+*/
+func getPinAddressOffset(pinNumber int) (int64, error) {
+
+	bankHeaderSize := 8 // 8 bytes of either header data assosciated with a bank (unsure about what is stored here though)
+
+	if !(1 <= pinNumber && pinNumber <= maxPinNum) {
+		return -1, errors.New("pin is out of bank range")
+	}
+
+	for i := 0; i < len(bankDivisions)-1; i++ {
+		if bankDivisions[i] <= pinNumber && pinNumber < bankDivisions[i+1] {
+
+			bankNum := i
+			bankBaseAddr := fselBankOffsets[bankNum]
+			pinBankOffset := pinNumber - bankDivisions[i]
+
+			pinAddressOffset := bankBaseAddr + bankHeaderSize + ((pinBankOffset) * fselPinDataSize)
+			return int64(pinAddressOffset), nil
+		}
+	}
+
+	return -1, errors.New("pin in bank range but not set")
+}
+
+// This method updates the given mode of a pin by finding its specific location in memory & writing to the 'mode' byte in the 8 byte block of pin data.
+func (b *pinctrlpi5) setPin(pinNumber int, newMode byte) error {
+
+	// Of the 8 bytes that represent a given pin's data, only the 4th index corresponds to the alternative mode setting
+	altModeIndex := 4
+
+	pinAddressOffset, err := getPinAddressOffset(pinNumber)
+	if err != nil {
+		return fmt.Errorf("error getting gpio bank number: %w", err)
+	}
+
+	// find pin data within virtual page; retrieve the 4th byte from the pin data
+	pinBytes := b.vPage[pinAddressOffset : pinAddressOffset+fselPinDataSize]
+	altModeByte := pinBytes[altModeIndex]
+
+	// We keep the left half of the byte preserved, only modifying the right half
+	// Preserve Left Side of Byte using this Mask
+	leftSideMask := byte(0xf0)
+
+	// Set Right Side of Byte ONLY using this Mask. Ensures left side cannot be overwritten
+	rightSideMask := byte(0x0f)
+
+	// Set new mode via write to correct while protecting previous other settings
+	newAltModeByte := (altModeByte & leftSideMask) | (newMode & rightSideMask)
+	pinBytes[4] = newAltModeByte
+
+	return err
+}
+
+// Sets up GPIO Pin Memory Access by parsing the device tree for relevant address information
+func (b *pinctrlpi5) setupPinControl() error {
+	nodePath, err := b.findPathFromAlias(gpioName) // this ("gpio") is hardcoded now, we will fix that later!
+	if err != nil {
+		b.logger.Errorf("error getting raspi5 GPIO nodePath")
+		return err
+	}
+
+	err = b.setGPIONodePhysAddr(nodePath)
+	if err != nil {
+		b.logger.Errorf("error getting raspi5 GPIO physical address")
+		return err
+	}
+
+	err = b.createGPIOVPage(gpioMemPath)
+	if err != nil {
+		b.logger.Errorf("error creating virtual page from GPIO physical address")
+		return err
+	}
+	return err
+}
+
+// Cleans up mapped memory / files related to pin control upon board close() call
+func (b *pinctrlpi5) cleanupPinControl() error {
```

Comment:
Would anything bad happen if we called `setupPinControl()` and then killed the process without calling `cleanupPinControl()` later? I suspect not, but want to double-check. I want to know if we're creating another problem like the libpigpio stuff...

Reply:
Based on the internet, the OS should clean up mapped files when processes are terminated, regardless of whether or not unmap is called. If the OS crashes for any reason, it also does memory cleanup as well upon reboot. This hasn't been tested though.

Reply:
after talking to @oliviamiller we're just going to leave this as is for now and not test it because everything should be fine. marking as resolved